### ISSUE TITLE
data_restore will fail when the length of Verified event is longer than Committed event

### ISSUE THREAD

**Issue by yaanhyy** [matter-labs/zksync#310]
https://github.com/matter-labs/zksync/blob/59e6def837099021169c35b7db7d7641f1a03b3b/core/bin/data_restore/src/events_state.rs#L364-L374

when   self.verified_events.len() >   self.committed_events.len()
        self.committed_events[0..count_to_get].to_vec() will overflow.

I meet this condition when restore databae.

I will give a bug fix.

**Comment by Deniallugo** [comment]
Hello! 
Thanks for your issue.
Could you provide more information about the situation when you met this problem? 
What was the environment?

**Comment by yaanhyy** [comment]
Sometimes,  http request is timeout. sometimes,  My mac is powerdown expected.

**Comment by Deniallugo** [comment]
After that failings, you tried to continue restoring with data-restorer flag `--continue`, am I right?
And it looks like the `continue` mode in the data restorer does not work correctly.

**Comment by yaanhyy** [comment]
yes, I use continue, I commit a fix on pull request and test it works. Please review:)

**Comment by StanislavBreadless** [comment]
Hey @yaanhyy!

We are very glad for your vigilance and support. But there can not be more verification events than commit events. Before each block verification, there is a corresponding block commit.

Could you please elaborate more on the steps to reproduce the issues? There are likely other problems that are the reason

**Comment by yaanhyy** [comment]
Sure I can reproduce the issues on follow steps:
1.  I run the data_restore service.
2. the  service  run update_operations_state and change data_restore_storage_state_update to operations
3. then service run update_tree_state , use the operations to produce blocks. when service produce some blocks but not finish. the service be shutdown for some unexpected reason(for example kill -9). 
4. then restart the data_restore service with --continue, the data_restore load  the storage_state_update in  operations state. 5. the service will load all  operations by function update_operations_state but some operations have been used to produce blocks already. then data_restore service will fail on update_tree_state.

**Comment by StanislavBreadless** [comment]
@yaanhyy 

We appreciate so many details. It seems that your PR has not resolved the issue because the reason for the error is that the program is trying to work with an inconsistent state. We have to look into our code and find where we are forgetting about a database transaction.
